domain/storage/service/storage_test.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/errors"
+	corestorage "github.com/juju/juju/core/storage"
+	domainstorage "github.com/juju/juju/domain/storage"
+	storageerrors "github.com/juju/juju/domain/storage/errors"
+	loggertesting "github.com/juju/juju/internal/logger/testing"
+	"github.com/juju/juju/internal/storage"
+	"github.com/juju/juju/internal/storage/provider"
+	coretesting "github.com/juju/juju/internal/testing"
+)
+
+type storageSuite struct {
+	testing.IsolationSuite
+
+	state              *MockState
+	registry           storage.ProviderRegistry
+	provider           *MockProvider
+	volumeSource       *MockVolumeSource
+	volumeImporter     *MockVolumeImporter
+	filesystemSource   *MockFilesystemSource
+	filesystemImporter *MockFilesystemImporter
+}
+
+var _ = gc.Suite(&storageSuite{})
+
+type volumeImporter struct {
+	*MockVolumeSource
+	*MockVolumeImporter
+}
+
+type filesystemImporter struct {
+	*MockFilesystemSource
+	*MockFilesystemImporter
+}
+
+func (s *storageSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.state = NewMockState(ctrl)
+	s.volumeSource = NewMockVolumeSource(ctrl)
+	s.volumeImporter = NewMockVolumeImporter(ctrl)
+	s.filesystemSource = NewMockFilesystemSource(ctrl)
+	s.filesystemImporter = NewMockFilesystemImporter(ctrl)
+	s.provider = NewMockProvider(ctrl)
+
+	registry := NewMockProviderRegistry(ctrl)
+	registry.EXPECT().StorageProvider(storage.ProviderType("ebs")).Return(s.provider, nil).AnyTimes()
+	registry.EXPECT().StorageProvider(storage.ProviderType("elastic")).Return(s.provider, nil).AnyTimes()
+
+	s.registry = storage.ChainedProviderRegistry{registry, provider.CommonStorageProviders()}
+
+	return ctrl
+}
+
+func (s *storageSuite) service(c *gc.C) *Service {
+	return NewService(s.state, loggertesting.WrapCheckLog(c), modelStorageRegistryGetter(func() storage.ProviderRegistry {
+		return s.registry
+	}))
+}
+
+func (s *storageSuite) TestImportFilesystemValidate(c *gc.C) {
```

Comment:
Ideally these could all just be `c.Check` and this could also be wrapped up into tests slice if you  wanted.

Reply:
There were only 3 of them so the overhead of a slice didn't seem worth it. Made them a Check